rfcs/convergence/first-rule-of-aria.md

```diff
@@ -0,0 +1,204 @@
+
+Some examples of where `<div>`s + ARIA fall down vs. semantic HTML include:
+
+- Some of the most popular voice control software has very limited support for ARIA
+- Windows High Contrast Mode only assigns color based on HTML semantics, not ARIA
+- Certain touch-based screen reader interactions do not trigger standard DOM events, so a control like `<input type="range">` is actually impossible to recreate accessibly with `<div>`s + ARIA (touch-based screen readers will not increment/decrement an ARIA slider)
+- Here's a thread that covers some of the differences between a `<textarea>` and a `<div role="textbox" contenteditable>`: https://twitter.com/codingchaos/status/1157001879991152640?s=20
+- A `<select>` element cannot be recreated with ARIA in a way that works seemlessly across platforms: mobile browsers often have completely custom rendering of the options list, and the platform API mappings are actually different between Windows, macOS, and iOS/Android.
+- Sometimes live regions are used to make up for differences between native behavior and custom ARIA widgets. This technique usually falls spectacularly flat on braille displays, where live regions have notably flaky support, particularly when they conflict with interaction.
+
+The difficulty replicating the robust nature of HTML with custom scripting and ARIA is why the [First Rule of ARIA](https://w3c.github.io/using-aria/#rule1) is "don't use ARIA".
+
+## Detailed Design or Proposal
+
+The proposal is a pattern in favor of using semantic elements as default for every convergence component. This is particularly important for interactive elements like buttons, links, inputs, radios, etc.
+
+The specific recommended steps to take are:
+
+- Whenever a semantic HTML element exists that covers the desired semantics and interaction, it should be used if at all possible.
+- Whenever we want to provide options beyond what HTML allows in an interactive widget, we should provide authors an option to use a semantic HTML version (e.g. Dropdown should have an author option to use a `<select>` element, even if we also allow a custom version with a stylable extendable options list).
+- When creating form components, use a `<label for="id">` element instead of `aria-label` or `aria-labelledby`.
+- When creating a component that does not exist in HTML like a tab control, try to make the base interactive controls using semantic HTML elements augmented with ARIA (so `<button role="tab">` instead of `<div role="tab">`) whenever it makes sense to do so.
```

Comment:
would this apply to menu items in a menu? In react-northstar, the menu item is an `<a />` and it brought some challenges with event handling. @miroslavstastny or @mshoho might remember the details.

Another cases where we came across challenges:
- List cannot be `<ul />` if you plan to include virtualization or custom scrollbar in it - as that will create `div`s, leading to invalid DOM structure
- Tables had styling challenges with native `<table />` - if I remember correctly that was related to column resize, @layershifter do you remember details?

Reply:
I think a good idea for components that simply can't follow this approach is to document it in the spec of the component expliciting the use cases that make this unviable, just to ensure that no one else in the future tries to implement this by mistake.

Reply:
There are two challenges with tables:
- Any virtualization library will break markup as `div`s will be inserted into a tree:

```tsx
<table>
  <tbody>
    {/* 👇 This markup is invalid and React will throw a warning */}
    <div>{/* nested elements */}</div>
  </tbody>
</table>
```

- I had serious troubles trying to implement resizing for tables (_I don't remember all details_), I am not sure that it's doable at all with table's tags. To clarify, I am referring to this scenario: https://react-table.tanstack.com/docs/examples/column-resizing

Reply:
Tables/grids are definitely one case where we might not want to use the native elements, for the two reasons @layershifter mentioned. I've also heard that browser performance for frequent DOM mutations in a large `<table>` can be poor, but I haven't tested that myself.

Table's a bit of a special case overall 😄. I also think generally this is much more important for interactive elements rather than layout elements.

Re: the `<a>` thing: link tags are super powerful when they have a meaningful `href`, but not really appropriate if they don't (and if the `href` is left off entirely, they are functionally equivalent to a `<span>` a11y-wise). For menu items, I think `<button>` would be a much better base.

Reply:
@smhigley interesting to use `button` as a base for `menuitems` since you'd be overriding the role in this case anyway, is it just to get the default space/enter behaviour ?